### PR TITLE
Add more permissive interface for RelationshipMeta options

### PR DESCRIPTION
**Pull request by mfeckie**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: Unfortunately  there's no simple link  to provide for this

Currently `RelationshipMeta` options  are  types as `object`.  This is incorrect in  practice.

Generally this  is only of interest to folks  creating custom adapters  and serializers, so not a huge deal, but when working on libraries (such as active-model-adapter), this is more troublesome.

This PR makes the options more permissive, documenting know options keys and providing an 'any' option for keys that are as yet unknown.

### CODE REVIEW

File: types/ember-data/index.d.ts
Comment (mike-north) [inline types/ember-data/index.d.ts]: One of the requirements of these types is that they describe only the library's public API. 

Looking through the docs for examples of this `RelationshipMeta` object, I can find [references to `async` and `inverse`](https://api.emberjs.com/ember-data/3.13/functions/@ember-data%2Fmodel/hasMany) but can't find anything that declares `polymorphic` to be a public thing. 

Do you have some basis for believing `polymorphic` is a public thing?

**Review by mike-north — CHANGES_REQUESTED** [review]

**Comment by mfeckie** [comment]
@mike-north The docs reference the property https://guides.emberjs.com/release/models/relationships/#toc_polymorphism when setting polymorphic relationships.

It's only really of interest to folks writing custom serializers, like this https://github.com/ember-data/active-model-adapter/blob/master/addon/active-model-serializer.js#L275-L281

**Comment by mike-north** [comment]
> @mike-north The docs reference the property https://guides.emberjs.com/release/models/relationships/#toc_polymorphism when setting polymorphic relationships

Historically, we've always used the API docs as the source of truth for what is/isn't in Ember or Ember-Data's public API. Please open up a PR adding the `polymorphic` property to the `belongsTo` and `hasMany` descriptions of allowed options. I'm happy to take an approval from an ember/ember-data core team member on that PR as confirmation.

**Comment by mfeckie** [comment]
OK, have opened PR to ember-data

**Comment by mfeckie** [comment]
@mike-north PR to ember-data has been accepted https://github.com/emberjs/data/pull/6593.  Could this proceed now?

**Review by chriskrycho — APPROVED** [review]
Thanks for doing that! You’re good to go now! (If DT accepts me; it may not—am on my phone and don’t recall if I’m on the reviewers list here.)

**Review by mike-north — APPROVED** [review]
Awesome, thanks!

**Pull request merged**
